pkg/cli/clusteroperator/restartkubelet/restart_kubelet.go

```diff
@@ -0,0 +1,229 @@
+package restartkubelet
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/oc/pkg/helpers/conditions"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	watchtools "k8s.io/client-go/tools/watch"
+)
+
+var (
+	nodeKind = schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Node"}
+
+	//go:embed restart-namespace.yaml
+	namespaceYaml []byte
+	namespace     = resourceread.ReadNamespaceV1OrDie(namespaceYaml)
+
+	//go:embed restart-pod-template.yaml
+	podYaml []byte
+	pod     = resourceread.ReadPodV1OrDie(podYaml)
+)
+
+type RestartKubeletRuntime struct {
+	ResourceFinder genericclioptions.ResourceFinder
+	KubeClient     kubernetes.Interface
+
+	DryRun bool
+
+	ImagePullSpec            string
+	NumberOfNodesInParallel  int
+	PercentOfNodesInParallel int
+
+	Printer printers.ResourcePrinter
+	genericclioptions.IOStreams
+}
+
+func (o *RestartKubeletRuntime) Run(ctx context.Context) error {
+	interestingNodes, err := o.GetInterestingNodes(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to get interesting nodes: %w", err)
+	}
+
+	numberOfNodesInParallel := 1
+	switch {
+	case o.NumberOfNodesInParallel > 0:
+		numberOfNodesInParallel = o.NumberOfNodesInParallel
+	case o.PercentOfNodesInParallel > 0:
+		numberOfNodesInParallel = len(interestingNodes) / o.PercentOfNodesInParallel
+	}
+	if numberOfNodesInParallel < 1 {
+		numberOfNodesInParallel = 1
+	}
+
+	// create a namespace to work in
+	nsName := "!!-dry-run"
+	if !o.DryRun {
+		ns, err := o.KubeClient.CoreV1().Namespaces().Create(ctx, namespace, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to create namespace: %w", err)
+		}
+		nsName = ns.Name
+		fmt.Fprintf(o.Out, "Created namespace/%v\n", ns.Name)
+		defer func() {
+			// using new context so the cancel doesn't stop the cleanup.
```

Comment:
Yeah but we don't hook up `oc` to SIGINT, so if the user Ctrl-C's the cli while waiting for kubelet restart, we will leak the namespace and also error out on later attempts rerun this command because of the unconditional `Create` above right?

Reply:
> Yeah but we don't hook up oc to SIGINT, so if the user Ctrl-C's the cli while waiting for kubelet restart, we will leak the namespace and also error out on later attempts rerun this command because of the unconditional Create above right?

Which signals should I catch?  I'm happy to update.

Reply:
Trying to properly handle SIGINT is hard.  And in the truly general case there's also Go panics (between the creation and `defer` setup), network flakes, operating system crashes - all of which can result in the cleanup code not being run.

Kubernetes wants us to do eventual consistency/reconcilation, so let's do that instead of trying to crash recovery?

i.e. idempotently create namespace, don't error if it exists, be sure we handle any pods created before etc

Reply:
> Kubernetes wants us to do eventual consistency/reconcilation, so let's do that instead of trying to crash recovery?

how would you do this with a client command as opposed to a long running process?

Reply:
I'm not blocking this PR to be clear.  If the process gets Ctrl-C'd, I guess we can leave it up to the user to deal with.  It looks like what would happen here is:

```
$ oc adm ocp-certificates restart-kubelet
Created namespace/openshift-restart-kubelet-xyz
# Ctrl-c or pod/machine/laptop dies or whatever
$ oc adm ocp-certificates  restart-kubelet
error: namespace openshift-restart-kubelet-xyz already exists
```
Then we'd just tell them to do

```
$ oc delete namespace/openshift-restart-kubelet-xyz
```
and restart?

Reply:
it's a generated namespace name.  We'd just get a new namespace when the command runs.